### PR TITLE
Support fixed_size_list for make_array

### PR DESCRIPTION
**Pull request by jayzhan211**
# Which issue does this PR close?

Ref https://github.com/apache/arrow-datafusion/issues/6560

# Rationale for this change

Add FixedSizeList support for array methods

# What changes are included in this PR?

Only MakeArray is done for this PR.

1. Add casting from fixedSizeList to list in analyzer (type_coercion).
2. Support ScalarValue::FixedSizeList, not all feature is implemented
3. Able to parse List for sqllogictest.

# Are these changes tested?

1. sqllogictest
array.slt

2. unit test in type_coercion.rs

# Are there any user-facing changes?

# Task
- [X] MakeArray
- [ ] Other array methods

### CODE REVIEW

File: datafusion/core/tests/data/fixed_size_list_array.parquet
Comment (jayzhan211) [inline datafusion/core/tests/data/fixed_size_list_array.parquet:1]: I create fixedsizelist parquet from my own script.
Is there any existing way for us to easily create/update this kind of test case?

File: Cargo.toml [hunk 1]
```diff
@@ -53,6 +53,14 @@ arrow-array = { version = "42.0.0", default-features = false, features = ["chron
 parquet = { version = "42.0.0", features = ["arrow", "async", "object_store"] }
 sqlparser = { version = "0.35", features = ["visitor"] }
 
+[patch.crates-io]
```
Comment (alamb) [inline Cargo.toml:56]: I think we will have to wait for this code to be released to crates.io (in about 10 days) before merging this PR

File: datafusion/common/src/scalar.rs [hunk 2]
```diff
@@ -3646,29 +3675,9 @@ impl fmt::Display for ScalarValue {
             ScalarValue::TimestampNanosecond(e, _) => format_option!(f, e)?,
             ScalarValue::Utf8(e) => format_option!(f, e)?,
             ScalarValue::LargeUtf8(e) => format_option!(f, e)?,
-            ScalarValue::Binary(e) => match e {
```
Comment (alamb) [inline datafusion/common/src/scalar.rs:3649]: ❤️

File: datafusion/core/tests/data/fixed_size_list_array.parquet
Comment (alamb) [reply datafusion/core/tests/data/fixed_size_list_array.parquet:1]: Did you look at the data sets available in https://github.com/apache/parquet-testing (which is a submodule of this repo)

e.g. https://github.com/apache/parquet-testing/tree/a11fc8f148f8a7a89d9281cc0da3eb9d56095fbf/data

File: datafusion/core/tests/sqllogictests/test_files/array.slt [hunk 3]
```diff
@@ -357,3 +357,33 @@ query ?
 select make_array(x, y) from foo2;
 ----
 [1.0, 1]
+
+statement ok
+CREATE EXTERNAL TABLE fixed_size_list_array STORED AS PARQUET LOCATION 'tests/data/fixed_size_list_array.parquet';
+
+query ?
+select * from fixed_size_list_array;
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/array.slt:365]: Could you also add a query that is like `arrow_typeof(f0)` to show in the test what the type of the column is?

File: datafusion/sql/src/expr/arrow_cast.rs [hunk 4]
```diff
@@ -486,6 +497,8 @@ impl<'a> Tokenizer<'a> {
             "Date32" => Token::SimpleType(DataType::Date32),
             "Date64" => Token::SimpleType(DataType::Date64),
 
+            "List" => Token::List,
```
Comment (alamb) [inline datafusion/sql/src/expr/arrow_cast.rs:500]: 👍

**Review by alamb — APPROVED** [review]
Thank you for this contribution @jayzhan211  -- I think other than the patch to crates.io this PR is ready to go. 
 
I left some suggestions, which I think would improve things

**Comment by alamb** [comment]
FYI @izveigor

**Comment by alamb** [comment]
Marking as a draft as it looks like this PR is not waiting on review -- it is waiting on an upstream release (scheduled early next week) as well as to address some other comments

File: datafusion/physical-expr/src/array_expressions.rs [hunk 5]
```diff
@@ -125,7 +125,7 @@ fn array_array(args: &[ArrayRef], data_type: DataType) -> Result<ArrayRef> {
             }
 
             let list_data_type =
-                DataType::List(Arc::new(Field::new("item", data_type, false)));
+                DataType::List(Arc::new(Field::new("item", data_type, true)));
```
Comment (jayzhan211) [inline datafusion/physical-expr/src/array_expressions.rs:128]: Revert the change from #666

**Review by alamb — APPROVED** [review]
Thank you @jayzhan211

**Comment by alamb** [comment]
cc @izveigor

**Pull request merged**